# Clean Transform MBS Subsidised Services & Benefits Data

In [130]:
# import libraries
import pandas as pd
import numpy as np
import os

## 1. MBS 2021-2022 Data Cleaning & Tranformation

In [131]:
# import the mbs data files and assign to a dataframe

# setup path to original dataset of mbs fileds
path = r'/Users/patel/Documents/CF-Data Anaylst Course/portfolio_projects/mbs_analysis/datasets/'

#import required columns
filter_cols = ["Worksheet","Year","StateTerritory","GeographicUnit","GeographicCode","GeographicAreaName","GeographicGroup", "ServiceLevel", "Service", "DemographicGroup","MeasureName","MeasureValue","Flag","FlagExplanation"]

df_mbs_2021 = pd.read_csv(os.path.join(path, 'original_datasets/mbs_data/phc-mbs-2021-2022/AIHW_MBS_2122_CSV.csv'), encoding='ISO-8859-1', usecols=filter_cols, index_col=None)
df_mbs_2021.head(3)

/var/folders/6w/mdyngzzd2_z7brmk404qky_m0000gs/T/ipykernel_31094/485236433.py:9: DtypeWarning: Columns (7) have mixed types. Specify dtype option on import or set low_memory=False.
  df_mbs_2021 = pd.read_csv(os.path.join(path, 'original_datasets/mbs_data/phc-mbs-2021-2022/AIHW_MBS_2122_CSV.csv'), encoding='ISO-8859-1', usecols=filter_cols, index_col=None)


Worksheet     Year StateTerritory GeographicUnit GeographicCode  \
0       PHN  2021-22       National            NAT         001NAT   
1       PHN  2021-22       National            NAT         001NAT   
2       PHN  2021-22       National            NAT         001NAT   

  GeographicAreaName GeographicGroup ServiceLevel                 Service  \
0           National        National      Level 1  GP attendances (total)   
1           National        National      Level 1  GP attendances (total)   
2           National        National      Level 1  GP attendances (total)   

  DemographicGroup                                   MeasureName  \
0             0-14  Percentage of people who had the service (%)   
1             0-14                       Services per 100 people   
2             0-14              Medicare benefits per 100 people   

   MeasureValue Flag FlagExplanation  
0       84.14%   NaN             NaN  
1        395.16  NaN             NaN  
2     17,792.31  NaN             NaN

In [132]:
df_mbs_2021.shape

(364176, 14)

In [133]:
df_mbs_2021.index.name = 'id'
df_mbs_2021.head()

Worksheet     Year StateTerritory GeographicUnit GeographicCode  \
id                                                                   
0        PHN  2021-22       National            NAT         001NAT   
1        PHN  2021-22       National            NAT         001NAT   
2        PHN  2021-22       National            NAT         001NAT   
3        PHN  2021-22       National            NAT         001NAT   
4        PHN  2021-22       National            NAT         001NAT   

   GeographicAreaName GeographicGroup ServiceLevel                 Service  \
id                                                                           
0            National        National      Level 1  GP attendances (total)   
1            National        National      Level 1  GP attendances (total)   
2            National        National      Level 1  GP attendances (total)   
3            National        National      Level 1  GP attendances (total)   
4            National        National      Level 1  GP attendances (total)   

   DemographicGroup                                   MeasureName  \
id                                                                  
0              0-14  Percentage of people who had the service (%)   
1              0-14                       Services per 100 people   
2              0-14              Medicare benefits per 100 people   
3              0-14                               No. of patients   
4              0-14                               No. of services   

    MeasureValue Flag FlagExplanation  
id                                     
0        84.14%   NaN             NaN  
1         395.16  NaN             NaN  
2      17,792.31  NaN             NaN  
3      3,997,279  NaN             NaN  
4     18,772,343  NaN             NaN

### 1.2 Extract and transform SA3 Worksheet Data

SA3 worksheet contains Medicare subsidsed servicees data by Statistical Area Level 3 (geograpghical area) 

In [134]:
# check the worksheet names to find SA3 reference
df_mbs_2021['Worksheet'].unique()

array(['PHN', 'SA3', 'PHN - Age standardised', 'PHN - GP aged care'],
      dtype=object)

In [135]:
df_mbs_2021[df_mbs_2021['Worksheet']=='SA3'].count()

Worksheet             231712
Year                  231712
StateTerritory        227680
GeographicUnit        231712
GeographicCode        231712
GeographicAreaName    231712
GeographicGroup       231712
ServiceLevel          231712
Service               231712
DemographicGroup      231712
MeasureName           231712
MeasureValue          231712
Flag                   16607
FlagExplanation        16607
dtype: int64

In [136]:
# extract SA3 data into a new dataframe
df_mbs_2021_sa3 = df_mbs_2021.loc[df_mbs_2021['Worksheet']=='SA3']
df_mbs_2021_sa3.head()

Worksheet     Year StateTerritory GeographicUnit GeographicCode  \
id                                                                       
132192       SA3  2021-22       National            NAT         001NAT   
132193       SA3  2021-22       National            NAT         001NAT   
132194       SA3  2021-22       National            NAT         001NAT   
132195       SA3  2021-22       National            NAT         001NAT   
132196       SA3  2021-22       National            NAT         001NAT   

       GeographicAreaName GeographicGroup ServiceLevel  \
id                                                       
132192           National        National      Level 1   
132193           National        National      Level 1   
132194           National        National      Level 1   
132195           National        National      Level 1   
132196           National        National      Level 1   

                       Service DemographicGroup  \
id                                                
132192  GP attendances (total)             0-24   
132193  GP attendances (total)             0-24   
132194  GP attendances (total)             0-24   
132195  GP attendances (total)             0-24   
132196  GP attendances (total)             0-24   

                                         MeasureName  MeasureValue Flag  \
id                                                                        
132192  Percentage of people who had the service (%)       83.76%   NaN   
132193                       Services per 100 people        435.77  NaN   
132194              Medicare benefits per 100 people     19,788.53  NaN   
132195                               No. of patients     6,578,439  NaN   
132196                               No. of services    34,223,467  NaN   

       FlagExplanation  
id                      
132192             NaN  
132193             NaN  
132194             NaN  
132195             NaN  
132196             NaN

In [137]:
print(df_mbs_2021_sa3.index.name)
df_mbs_2021_sa3.shape

id


(231712, 14)

In [138]:
df_mbs_2021_sa3.isna().sum()

Worksheet                  0
Year                       0
StateTerritory          4032
GeographicUnit             0
GeographicCode             0
GeographicAreaName         0
GeographicGroup            0
ServiceLevel               0
Service                    0
DemographicGroup           0
MeasureName                0
MeasureValue               0
Flag                  215105
FlagExplanation       215105
dtype: int64

In [139]:
df_mbs_2021_sa3[df_mbs_2021_sa3['StateTerritory'].isna()]

Worksheet     Year StateTerritory GeographicUnit GeographicCode  \
id                                                                       
132864       SA3  2021-22            NaN            SA3         004-01   
132865       SA3  2021-22            NaN            SA3         004-01   
132866       SA3  2021-22            NaN            SA3         004-01   
132867       SA3  2021-22            NaN            SA3         004-01   
132868       SA3  2021-22            NaN            SA3         004-01   
...          ...      ...            ...            ...            ...   
136891       SA3  2021-22            NaN            SA3         004-06   
136892       SA3  2021-22            NaN            SA3         004-06   
136893       SA3  2021-22            NaN            SA3         004-06   
136894       SA3  2021-22            NaN            SA3         004-06   
136895       SA3  2021-22            NaN            SA3         004-06   

                GeographicAreaName             GeographicGroup ServiceLevel  \
id                                                                            
132864   Major cities - higher SES   Major cities - higher SES      Level 1   
132865   Major cities - higher SES   Major cities - higher SES      Level 1   
132866   Major cities - higher SES   Major cities - higher SES      Level 1   
132867   Major cities - higher SES   Major cities - higher SES      Level 1   
132868   Major cities - higher SES   Major cities - higher SES      Level 1   
...                            ...                         ...          ...   
136891  Remote (incl. very remote)  Remote (incl. very remote)      Level 3   
136892  Remote (incl. very remote)  Remote (incl. very remote)      Level 3   
136893  Remote (incl. very remote)  Remote (incl. very remote)      Level 3   
136894  Remote (incl. very remote)  Remote (incl. very remote)      Level 3   
136895  Remote (incl. very remote)  Remote (incl. very remote)      Level 3   

                       Service DemographicGroup  \
id                                                
132864  GP attendances (total)             0-24   
132865  GP attendances (total)             0-24   
132866  GP attendances (total)             0-24   
132867  GP attendances (total)             0-24   
132868  GP attendances (total)             0-24   
...                        ...              ...   
136891        Speech Pathology      All persons   
136892        Speech Pathology      All persons   
136893        Speech Pathology      All persons   
136894        Speech Pathology      All persons   
136895        Speech Pathology      All persons   

                                         MeasureName      MeasureValue Flag  \
id                                                                            
132864  Percentage of people who had the service (%)           81.88%   NaN   
132865                       Services per 100 people            434.05  NaN   
132866              Medicare benefits per 100 people         19,555.73  NaN   
132867                               No. of patients           979,505  NaN   
132868                               No. of services         5,192,202  NaN   
...                                              ...               ...  ...   
136891                               No. of patients               181  NaN   
136892                               No. of services               460  NaN   
136893                  Total Medicare benefits paid           $26,965  NaN   
136894                           Total provider fees           $50,679  NaN   
136895                 Estimated resident population            535932  NaN   

       FlagExplanation  
id                      
132864             NaN  
132865             NaN  
132866             NaN  
132867             NaN  
132868             NaN  
...                ...  
136891             NaN  
136892             NaN  
136893             NaN  
136894             NaN  
136895             NaN  

State Terriory had 4,032 blanls because SA3 Group aggregages are not for a specific state or terriorty. They are showing aggregrates of each SA3 group by demographics and service.

SA3 groups:
004-01 = Major cities - higher SES
004-02 = Major cities - medium SES
004-03 = Major cities - lower SES
004-04 = Inner regional
004-05 = Outer regional
004-06 = Remote (incl. very remote)

In [140]:
# to ensure pivot table does not skip the blank stateterritory, filling in the blank cells.
df_mbs_2021_sa3_fill_na = df_mbs_2021_sa3.copy()
df_mbs_2021_sa3_fill_na['StateTerritory'].fillna('National-SA3Group', inplace=True)

In [141]:
print(df_mbs_2021_sa3_fill_na.isna().sum())
df_mbs_2021_sa3_fill_na.shape

Worksheet                  0
Year                       0
StateTerritory             0
GeographicUnit             0
GeographicCode             0
GeographicAreaName         0
GeographicGroup            0
ServiceLevel               0
Service                    0
DemographicGroup           0
MeasureName                0
MeasureValue               0
Flag                  215105
FlagExplanation       215105
dtype: int64


(231712, 14)

In [142]:
# Use pivot_table to reshape the DataFrame
df_mbs_2021_sa3_pivot = df_mbs_2021_sa3_fill_na.pivot_table(index=['Worksheet', 'Year', 'StateTerritory', 'GeographicUnit', 'GeographicCode', 
                                                           'GeographicAreaName', 'GeographicGroup', 'ServiceLevel', 'Service', 'DemographicGroup'],
                            columns='MeasureName', 
                            values='MeasureValue',
                            aggfunc='first')  # 'first' is used to pick the first value in case of duplicates

# Reset index to make the DataFrame flat
df_mbs_2021_sa3_pivot.reset_index(inplace=True)
df_mbs_2021_sa3_pivot.head(5)

MeasureName Worksheet     Year StateTerritory GeographicUnit GeographicCode  \
0                 SA3  2021-22            ACT            SA3          80101   
1                 SA3  2021-22            ACT            SA3          80101   
2                 SA3  2021-22            ACT            SA3          80101   
3                 SA3  2021-22            ACT            SA3          80101   
4                 SA3  2021-22            ACT            SA3          80101   

MeasureName GeographicAreaName            GeographicGroup ServiceLevel  \
0                    Belconnen  Major cities - medium SES      Level 1   
1                    Belconnen  Major cities - medium SES      Level 1   
2                    Belconnen  Major cities - medium SES      Level 1   
3                    Belconnen  Major cities - medium SES      Level 1   
4                    Belconnen  Major cities - medium SES      Level 1   

MeasureName                            Service DemographicGroup  \
0            Allied Health attendances (total)             0-24   
1            Allied Health attendances (total)            25-44   
2            Allied Health attendances (total)            45-64   
3            Allied Health attendances (total)              65+   
4            Allied Health attendances (total)      All persons   

MeasureName Estimated resident population Medicare benefits per 100 people  \
0                                   33439                         4,385.95   
1                                   34495                         6,177.09   
2                                   22397                         6,283.74   
3                                   15541                         7,940.67   
4                                  105872                         5,892.81   

MeasureName No. of patients No. of services  \
0                     6,371          16,206   
1                     8,921          23,460   
2                     9,848          19,883   
3                     9,794          21,140   
4                    34,933          80,690   

MeasureName Percentage of people who had the service (%)  \
0                                                19.05%    
1                                                25.86%    
2                                                43.97%    
3                                                63.02%    
4                                                33.00%    

MeasureName Services per 100 people Total Medicare benefits paid  \
0                             48.46                   $1,466,617   
1                             68.01                   $2,130,787   
2                             88.78                   $1,407,369   
3                            136.03                   $1,234,059   
4                             76.21                   $6,238,832   

MeasureName Total provider fees  
0                    $2,259,631  
1                    $3,248,640  
2                    $1,933,615  
3                    $1,594,820  
4                    $9,036,706

In [143]:
df_mbs_2021_sa3_pivot.isna().sum()

MeasureName
Worksheet                                       0
Year                                            0
StateTerritory                                  0
GeographicUnit                                  0
GeographicCode                                  0
GeographicAreaName                              0
GeographicGroup                                 0
ServiceLevel                                    0
Service                                         0
DemographicGroup                                0
Estimated resident population                   0
Medicare benefits per 100 people                0
No. of patients                                 0
No. of services                                 0
Percentage of people who had the service (%)    0
Services per 100 people                         0
Total Medicare benefits paid                    0
Total provider fees                             0
dtype: int64

In [144]:
#removed the column index 'MeasureValue' to simply the table
df_mbs_2021_sa3_pivot.columns.name = None
df_mbs_2021_sa3_pivot.head()

Worksheet     Year StateTerritory GeographicUnit GeographicCode  \
0       SA3  2021-22            ACT            SA3          80101   
1       SA3  2021-22            ACT            SA3          80101   
2       SA3  2021-22            ACT            SA3          80101   
3       SA3  2021-22            ACT            SA3          80101   
4       SA3  2021-22            ACT            SA3          80101   

  GeographicAreaName            GeographicGroup ServiceLevel  \
0          Belconnen  Major cities - medium SES      Level 1   
1          Belconnen  Major cities - medium SES      Level 1   
2          Belconnen  Major cities - medium SES      Level 1   
3          Belconnen  Major cities - medium SES      Level 1   
4          Belconnen  Major cities - medium SES      Level 1   

                             Service DemographicGroup  \
0  Allied Health attendances (total)             0-24   
1  Allied Health attendances (total)            25-44   
2  Allied Health attendances (total)            45-64   
3  Allied Health attendances (total)              65+   
4  Allied Health attendances (total)      All persons   

  Estimated resident population Medicare benefits per 100 people  \
0                         33439                         4,385.95   
1                         34495                         6,177.09   
2                         22397                         6,283.74   
3                         15541                         7,940.67   
4                        105872                         5,892.81   

  No. of patients No. of services  \
0           6,371          16,206   
1           8,921          23,460   
2           9,848          19,883   
3           9,794          21,140   
4          34,933          80,690   

  Percentage of people who had the service (%) Services per 100 people  \
0                                      19.05%                    48.46   
1                                      25.86%                    68.01   
2                                      43.97%                    88.78   
3                                      63.02%                   136.03   
4                                      33.00%                    76.21   

  Total Medicare benefits paid Total provider fees  
0                   $1,466,617          $2,259,631  
1                   $2,130,787          $3,248,640  
2                   $1,407,369          $1,933,615  
3                   $1,234,059          $1,594,820  
4                   $6,238,832          $9,036,706

In [145]:
# check number of rows and columns as expected
df_mbs_2021_sa3_pivot.shape

(28964, 18)

In [146]:
# counting distinct values. Checking for 'National-SA3Group'
df_mbs_2021_sa3_pivot['StateTerritory'].value_counts()

NSW                  7623
Qld                  6887
Vic                  5543
WA                   2847
SA                   2350
Tas                  1257
ACT                   840
NT                    750
National-SA3Group     504
Other Territories     279
National               84
Name: StateTerritory, dtype: int64

In [147]:
df_mbs_2021_sa3_pivot.to_csv(os.path.join(path, 'clean_datasets/mbs_data/2021-22_phc_mbs.csv'))